File: JAVA/112 - Path Sum.java
/*
 112. Path Sum (E)

 Approach 1 : 
 Time: O(n)
 Space: O(1)

 Approach 2 : DFS 
 Time: O(n)
 Space: O(2n)

*/

/* Approach 1 */
class Solution {
    public boolean hasPathSum(TreeNode root, int sum) {
        if (root == null) return false;
        else if (root.val == sum && root.left == null && root.right ==null)
            return true;
        else 
            return hasPathSum(root.left, sum - root.val) || hasPathSum(root.right, sum - root.val);
    }
}

/* Approach 2 */
class Solution {
    public boolean hasPathSum(TreeNode root, int sum) {
        if (root == null) return false;
        Stack<TreeNode> todo = new Stack<TreeNode>();
        Stack<Integer> sumStack = new Stack<Integer>();
        todo.push(root);
        sumStack.push(root.val);
        while (!todo.empty()) {
            root = todo.pop();
            int sumTmp = sumStack.pop();
            
            if (root.left == null && root.right == null && sumTmp == sum)
                return true;
            
            if (root.right != null) {
                todo.push(root.right);
                sumStack.push(sumTmp + root.right.val);
            }
            if (root.left != null) {
                todo.push(root.left);
                sumStack.push(sumTmp + root.left.val);
            }
        }
        return false;
    }
}

File: C++/112-Path_Sum.cpp
/*
 112. Path Sum (E)

 Preorder(top-down)
 Preorder the tree, if the path sum isn't fullfill the given sum, move to its 
 child and add its value
 T:O(n)
 S:O(leaf)

 Preorder(top-down) recursion
 If the current node isn't a leaf, deduct sum with the current node value and
 move to its child 
 T:O(n)
 S:O(depth)
 
*/

// Preorder(top-down)
class Solution {
public:
    bool hasPathSum(TreeNode* root, int sum) {
        if (!root) return false;
        // Queue for todo nodes and current sum
        queue<TreeNode*> todo; 
        queue<int> sums;
        todo.push(root);
        sums.push(root->val);

        while (!todo.empty()) { 
            root = todo.front();
            todo.pop();
            int curSum = sums.front();
            sums.pop();
            
            if (!root->left && !root->right && curSum == sum) {
                return true;
            }
            if (root->left) {
                todo.push(root->left);
                sums.push(curSum + root->left->val);
            }
            if (root->right) {
                todo.push(root->right);
                sums.push(curSum + root->right->val);
            }
        }
        return false;
    }
};


// Preorder(top-down) recursion
class Solution {
public:
    bool hasPathSum(TreeNode* root, int sum) {
        // Use leaf to stop the recursion because we can't know the entry point
        if (!root) return false;
        if (!root->left && !root->right && root->val == sum) { 
            return true;
        }
        return hasPathSum(root->left, sum - root->val) 
            || hasPathSum(root->right, sum - root->val);
    }
};
